### PR TITLE
Introduce contract key to exercise node

### PR DESCRIPTION
**Pull request by dajmaki**
Track the contract key in NodeExercises. This is in
preparation for adding the key into exercise events, and
for support for contract keys in kvutils, where we need
to be able to compute the outputs of the submission, e.g.
the contract key that would be unset by a consuming exercise.

This introduces transaction version 8.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala [hunk 1]
```diff
@@ -205,12 +209,14 @@ object Node {
...
-            exerciseResult2) =>
+            exerciseResult2,
+            key2) =>
           import ne._
           targetCoid === targetCoid2 && templateId == templateId2 && choiceId == choiceId2 &&
           consuming == consuming2 && actingParties == actingParties2 && chosenValue === chosenValue2 &&
           stakeholders == stakeholders2 && signatories == signatories2 && controllers == controllers2 &&
-          exerciseResult.fold(true)(_ => exerciseResult === exerciseResult2)
+          exerciseResult.fold(true)(_ => exerciseResult === exerciseResult2) &&
+          key === key2
```
Comment (S11001001) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala]: If validating a past transaction (where key = None) will produce a NodeExercises where key = Some, then this needs to use the same approach as `exerciseResult`.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/VersionTimeline.scala [hunk 2]
```diff
@@ -53,6 +53,7 @@ private[lf] object VersionTimeline {
       That(LanguageVersion(LMV.V1, "4")),
       That(LanguageVersion(LMV.V1, "5")),
       That(LanguageVersion(LMV.V1, Dev)),
+      This(That(TransactionVersion("8"))),
```
Comment (S11001001) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/VersionTimeline.scala]: Let's put the 8 above Dev; see the comment below that addresses why dev versions are weird.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionCoderSpec.scala [hunk 3]
```diff
@@ -172,10 +179,26 @@ class TransactionCoderSpec
...
+                                                   transactionWithout(decWithMax.transaction, {
+                                                     x: Tx.Node =>
+                                                       withoutExerciseResult(
+                                                         withoutContractKeyInExercise(x))
+                                                   })
+                                                 else if (txvMin precedes minContractKeyInExercise)
+                                                   transactionWithout(decWithMax.transaction, {
+                                                     x: Tx.Node =>
+                                                       withoutContractKeyInExercise(x)
+                                                   })
```
Comment (S11001001) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionCoderSpec.scala]: You're making me feel bad about introducing this pattern in the first place :)

Would you  mind factoring this pattern out even further? Taking the transaction and the txVer/txvMin.

File: unreleased.rst [hunk 4]
```diff
@@ -15,3 +15,10 @@ DAML Studio
 - The selected view for scenario results (table or transaction) is now
   preserved when the scenario results are updated.
   See `#1675 <https://github.com/digital-asset/daml/issues/1675>`__.
+
+DAML Integration Kit
+~~~~~~~~~~~~~~~~~~~~
+
+- Contract key, if any, is now included in the transaction exercise node. This prepares
+  the way for #1456 and #1680.
```
Comment (S11001001) [inline unreleased.rst]: I don't think this needs a release note at all, but if it does it should include a full URL as with the other notes.

File: daml-lf/transaction/src/main/protobuf/com/digitalasset/daml/lf/transaction.proto [hunk 5]
```diff
@@ -11,7 +11,8 @@
 // * 4: string contract_id replaced globally by ContractId message
 // * 5: new field actors in NodeFetch
 // * 6: removal of controllers in exercise nodes
-// * 7: new field exerciseResult in NodeExercise
+// * 7: new field return_value in NodeExercise
+// * 8: new field contract_key in NodeExercise
```
Comment (S11001001) [inline daml-lf/transaction/src/main/protobuf/com/digitalasset/daml/lf/transaction.proto]: Please update `daml-lf/spec/transaction.rst` describing the new semantics as well:

* [ ] version at top of file
* [ ] entry in version history table
* [ ] definition of the field at the end of the message NodeExercise section

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala (hunk 1 above)
Comment (dajmaki) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala]: thanks, good catch!

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionCoderSpec.scala (hunk 3 above)
Comment (S11001001) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionCoderSpec.scala]: Something like

```scala
  def minimalistTx[Nid, Cid, Val](txvMin: TransactionVersion, tx: GenTransaction[Nid, Cid, Val]): GenTransaction[Nid, Cid, Val] =
        if (txvMin precedes minExerciseResult)
                                                   transactionWithout(tx, { x: Tx.Node =>
                                                     withoutExerciseResult(
                                                       withoutContractKeyInExercise(x))
                                                   })
                                                 else if (txvMin precedes minContractKeyInExercise)
                                                   transactionWithout(tx, { x: Tx.Node =>
                                                     withoutContractKeyInExercise(x)
                                                   })
                                                 else tx
```

Comment (dajmaki) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionCoderSpec.scala]: thanks, done.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/VersionTimeline.scala (hunk 2 above)
Comment (dajmaki) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/VersionTimeline.scala]: moved.

File: unreleased.rst (hunk 4 above)
Comment (dajmaki) [reply unreleased.rst]: removed it.

File: daml-lf/transaction/src/main/protobuf/com/digitalasset/daml/lf/transaction.proto (hunk 5 above)
Comment (dajmaki) [reply daml-lf/transaction/src/main/protobuf/com/digitalasset/daml/lf/transaction.proto]: updated.

File: unreleased.rst [hunk 6]
```diff
@@ -8,3 +8,15 @@ This page contains release notes for the SDK.
...
+~~~~~~~~~~~
+
+- The selected view for scenario results (table or transaction) is now
+  preserved when the scenario results are updated.
+  See `#1675 <https://github.com/digital-asset/daml/issues/1675>`__.
+
+DAML Integration Kit
+~~~~~~~~~~~~~~~~~~~~
+
+- [DAML-LF] Rename ``none`` and ``some`` to ``optional_none`` and ``optional_some``, resp., in ``Expr`` and ``CasePat``.
```
Comment (S11001001) [inline unreleased.rst]: Reintroduced by conflict resolution?

**Pull request merged**